src/cplscheme/ParallelCouplingScheme.cpp

```diff
@@ -37,22 +37,43 @@ void ParallelCouplingScheme::exchangeFirstData()
 
 void ParallelCouplingScheme::exchangeSecondData()
 {
+  // @todo bundle receiveConvergence, specific moveToNextWindow and sendData into one function?
   if (doesFirstStep()) { // first participant
-    PRECICE_DEBUG("Receiving data...");
     if (isImplicitCouplingScheme()) {
+      PRECICE_DEBUG("Receiving convergence data...");
       receiveConvergence(getM2N());
     }
+    if (hasConverged() || isExplicitCouplingScheme()) {
+      moveToNextWindow();
+    }
+    PRECICE_DEBUG("Receiving data...");
     receiveData(getM2N(), getReceiveData());
     checkDataHasBeenReceived();
-  } else { // second participant
-    if (isImplicitCouplingScheme()) {
+  }
+
+  if (isImplicitCouplingScheme()) {
+    if (not doesFirstStep()) { // second participant
       PRECICE_DEBUG("Perform acceleration (only second participant)...");
       doImplicitStep();
+    }
+  }
+
+  // @todo bundle sendConvergence, specific moveToNextWindow and sendData into one function?
+  if (not doesFirstStep()) {
+    if (isImplicitCouplingScheme()) {
+      PRECICE_DEBUG("Sending convergence...");
       sendConvergence(getM2N());
     }
+    if (hasConverged() || isExplicitCouplingScheme()) {
+      moveToNextWindow();
+    }
     PRECICE_DEBUG("Sending data...");
     sendData(getM2N(), getSendData());
   }
+
+  if (isImplicitCouplingScheme()) {
+    storeIteration();
+  }
```

Comment:
Readability-wise, wouldn't it be simpler to restructure here into on big:
```cpp
if (isExplicitCouplingScheme()) {
...
} else {
  assert(isImplicitCouplingScheme());
 ...
}
```
(same in `SerialCouplingScheme` class)

Reply:
Not so sure. Changed it in e3c1f51484c2ca3817b98efc4021712b359d418d, we can still revert. But this way we at least have both variants to compare.

Reply:
Much easier to read IMO. One starts with the easy thing (explicit) and understands the basics and then moves to the complicated thing (implicit) and sees what else is still important. 
There are even more `isImplicitCouplingScheme()` you could remove.